src/CoreWCF.Primitives/src/CoreWCF/Xml/XmlDictionaryWriterExtensions.cs

```diff
@@ -0,0 +1,20 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Xml;
+
+namespace CoreWCF.Xml
+{
+    internal static class XmlDictionaryWriterExtensions
+    {
+        private static readonly bool s_frameworkSupportsAsyncWriteOperations = Environment.Version.Major >= 6;
+
+        internal static bool SupportsAsync(this XmlDictionaryWriter writer)
+        {
+            // Unfortunately not every XmlWriter supports asynchronous IO.
+            // And there's no API to check for this.
+            return s_frameworkSupportsAsyncWriteOperations && writer is IAsyncXmlWriter;
```

Comment:
There's a second check you can also add to this. You can check `writer.Settings?.IsAsync ?? false`. This should light up this code path for more scenarios and make more things async.

Reply:
You still need your IAsyncXmlWriter method because XmlMtomWriter.Settings returns the Settings property from the inner XmlWriter, which in the scenarios I was investigating was always null.